File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:     
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        
        def dfs(root, depth):
            if not root:
                return (depth-1, None)
            
            left = dfs(root.left,depth+1)
            right = dfs(root.right, depth+1)
            
            if left[0] == right[0]:
                return (left[0], root)
            if left[0] > right[0]:
                return left
            else:
                return right
            
        return dfs(root,0)[1]

File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<TreeNode*,int> dfs(TreeNode* root, int d) {
        if (!root)
            return {NULL, d - 1};
        if (!root->left && !root->right)
            return {root, d};
        auto left = dfs(root->left, d + 1);
        auto right = dfs(root->right, d + 1);
        if (left.second == right.second)
            return {root, left.second};
        else if (left.second < right.second)
            return right;
        else
            return left;
    }
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        return dfs(root, 0).first;
    }
};